src/Cardano/Shell/NodeIPC/Lib.hs

```diff
@@ -162,44 +176,58 @@ getIPCHandle = do
+startIPC :: forall m. (MonadIO m) => ProtocolDuration -> ReadHandle -> WriteHandle -> Port -> m ()
+startIPC protocolDuration readHandle writeHandle port = liftIO $ void $ ipcListener protocolDuration readHandle writeHandle port
 
 -- | Start IPC with NodeJS
 --
 -- This only works if NodeJS spawns the Haskell executable as child process
 -- (See @server.js@ as an example)
-startNodeJsIPC :: forall m. (MonadIO m) => Port -> m ()
-startNodeJsIPC port = do
-    handle <- liftIO $ getIPCHandle
-    let readHandle = ReadHandle handle
+startNodeJsIPC :: forall m. (MonadIO m) => ProtocolDuration -> Port -> m ()
+startNodeJsIPC protocolDuration port = do
+    handle          <- liftIO $ getIPCHandle
+    let readHandle  = ReadHandle handle
     let writeHandle = WriteHandle handle
-    liftIO $ void $ ipcListener readHandle writeHandle port
+    liftIO $ void $ ipcListener protocolDuration readHandle writeHandle port
+
+-- | Function for handling the protocol
+handleIPCProtocol :: forall m. (MonadIO m) => Port -> MsgIn -> m MsgOut
+handleIPCProtocol (Port port) = \case
+    QueryPort          -> pure (ReplyPort port)
+    Ping               -> pure Pong
+    MessageInFailure f -> pure $ MessageOutFailure f
 
 -- | Start IPC listener with given Handles and Port
 --
 -- When the listener recieves 'Ping' it will return 'Pong'.
 --
 -- If it recieves 'QueryPort', then the listener
 -- responds with 'ReplyPort' with 'Port',
-ipcListener :: forall m . (MonadIO m, MonadCatch m, MonadMask m) => ReadHandle -> WriteHandle -> Port -> m ()
-ipcListener readHandle@(ReadHandle rHndl) writeHandle@(WriteHandle wHndl) (Port port) =
-    do
-        checkHandles readHandle writeHandle
-        catches handleMsgIn [Handler handler, Handler handleMsgError]
-    `finally`
-    shutdown
+ipcListener
+    :: forall m. (MonadIO m, MonadCatch m, MonadMask m)
+    => ProtocolDuration
+    -> ReadHandle
+    -> WriteHandle
+    -> Port
+    -> m ()
+ipcListener protocolDuration readHandle@(ReadHandle rHndl) writeHandle@(WriteHandle wHndl) port = do
+    checkHandles readHandle writeHandle
+    handleMsgIn `catches` [Handler handler, Handler handleMsgError] `finally` shutdown
   where
     handleMsgIn :: m ()
     handleMsgIn = do
         liftIO $ hSetNewlineMode rHndl noNewlineTranslation
-        send Started
-        msgIn <- readMessage readHandle
-        case msgIn of
-            QueryPort          -> send (ReplyPort port)
-            Ping               -> send Pong
-            -- TODO:Handle them nicely
-            MessageInFailure _ -> return ()
+        send Started -- Send the message first time the IPC is up!
+
+        let frequencyFunction = case protocolDuration of
+                                    SingleMessage -> void
+                                    MultiMessage  -> forever
+
+        -- Fetch message and respond to it
+        frequencyFunction $ do
+            msgIn               <- readMessage readHandle        -- Read message
```

Comment:
The main reason why I can't generalize the function, we don't know what the type is.